src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -256,90 +267,120 @@ private static boolean shouldBeDeclaredAsFinal(ClassDesc desc) {
         if (superClassName != null) {
-            final ClassDesc nearest =
-                    getNearestClassWithSameName(superClassName, qualifiedClassName);
-            if (nearest == null) {
-                Optional.ofNullable(innerClasses.get(superClassName))
-                        .ifPresent(ClassDesc::registerNestedSubclass);
-            }
-            else {
-                nearest.registerNestedSubclass();
-            }
+            final Optional<ClassDesc> nearest =
+                getNearestClassWithSameName(superClassName,
+                                            className -> {
+                                                return CheckUtil.typeDeclarationNameMatchingCount(
+                                                    qualifiedClassName, className);
+                                            });
+            nearest.ifPresentOrElse(ClassDesc::registerNestedSubclass,
+                                    () -> {
+                                        Optional
+                                            .ofNullable(innerClasses.get(superClassName))
+                                            .ifPresent(ClassDesc::registerNestedSubclass);
+                                    });
         }
     }
 
     /**
-     * Checks if there is a class with same name.
+     * Register to the super class of anonymous inner class that the given class is instantiated
+     * by an anonymous inner class.
+     *
+     * @param literalNewAst ast node of {@link TokenTypes#LITERAL_NEW} representing anonymous inner
+     *                      class
+     * @param outerTypeDeclName Fully qualified name of the outer type declaration of anonymous
+     *                          inner class
+     */
+    private void registerAnonymousInnerClassToSuperClass(DetailAST literalNewAst,
+                                                         String outerTypeDeclName) {
+        final String superClassName = CheckUtil.getShortNameOfAnonInnerClass(literalNewAst);
+        final Optional<ClassDesc> superClassDesc =
+            getNearestClassWithSameName(superClassName,
+                                        className -> {
+                                            return CheckUtil
+                                                .getAnonSuperTypeDeclarationMatchingCount(
+                                                    outerTypeDeclName, className);
+                                        });
+        superClassDesc.ifPresentOrElse(ClassDesc::registerSuperClassOfAnonymousInnerClass,
+                                       () -> {
+                                           Optional.ofNullable(innerClasses.get(superClassName))
+                                               .ifPresent(
+                                                   ClassDesc
+                                                       ::registerSuperClassOfAnonymousInnerClass);
+                                       });
```

Comment:
I do not like this stairs-style, it distracts from following main flow of the logic. Please make it more flat (same for other method with stairs)

I can suggest following:
1. no need for intermediate optional variable. can be one flow
2. registerSuperClassOfAnonymousInnerClass and registerNestedSubclass can accept nulls, therefore eliminating need to make additional wrap into optional here
3. Try to make CheckUtil usage singleline. I guess NeedBraces rule forced you to add return and braces. So, name of the method can be shortened and lambda extracted as local variable

Reply:
To make it clearer, I resolve this and respond in separate comment